sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -198,7 +204,38 @@ case class TransformWithStateExec(
       getOutputRow(obj)
     }
     ImplicitGroupingKeyTracker.removeImplicitKey()
-    mappedIterator
+
+    // Wrapper to ensure that the implicit key is set when the methods on the iterator
+    // are called. Inside of processNewData, we use a GroupedIterator, so handleInputRows
+    // is only called once per key. As such, we only have to set the implicit key when
+    // the first call to hasNext is made, and we have to remove it when hasNext returns
+    // false.
+    //
+    // Note: if we ever start to interleave the processing of the iterators we get back
+    // from handleInputRows (i.e. we don't process each iterator all at once), then this
+    // iterator will need to set/unset the implicit key every time hasNext/next is called,
+    // not just at the first and last calls to hasNext.
+    new Iterator[InternalRow] {
+      var hasStarted = false
+
+      override def hasNext: Boolean = {
+        if (!hasStarted) {
+          hasStarted = true
+          ImplicitGroupingKeyTracker.setImplicitKey(keyObj)
+        }
+
+        val hasNext = mappedIterator.hasNext
+        if (!hasNext) {
+          ImplicitGroupingKeyTracker.removeImplicitKey()
+        }
+        hasNext
+      }
+
+      override def next(): InternalRow = {
+        assert(hasStarted, "next called before hasNext")
```

Comment:
Could we improve the error message ? maybe add info about the iterator/function call maybe ?

Reply:
Hm, but this is really an internal assertion—as developers, seeing this assertion on this line will be very clear as to what happened. I can improve it to be something like, "next was called on iterator for handleInputRows before hasNext" ?

Reply:
yea something like `Failed to invoke hasNext() before invoking next() while processing output iterator from handleInputRows` ?

Reply:
Done (for both).